common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java

```diff
@@ -821,20 +914,132 @@ public void testFailureAfterDuplicateBlockDoesNotInterfereActiveStream() throws
+      new PushBlockStream(testApp, 2, 0, 1, 0, 0));
+    for (ByteBuffer block : blocks) {
+      stream2.onData(stream2.getID(), block);
+    }
+    stream2.onComplete(stream2.getID());
+    closed.acquire();
+    // Check if all the file channels created for the first attempt are safely closed.
+    for (Map<Integer, RemoteBlockPushResolver.AppShufflePartitionInfo> partitionMap :
+        partitions.values()) {
+      for (RemoteBlockPushResolver.AppShufflePartitionInfo partitionInfo : partitionMap.values()) {
+        assertNull(partitionInfo.getDataChannel());
+        assertNull(partitionInfo.getMetaFile());
+        assertNull(partitionInfo.getIndexFile());
+      }
+    }
+    try {
+      pushResolver.receiveBlockDataAsStream(
+        new PushBlockStream(testApp, 1, 0, 1, 0, 0));
+    } catch (IllegalArgumentException re) {
+      assertEquals(
+        "The attempt id 1 in this PushBlockStream message does not match " +
+          "with the current attempt id 2 stored in shuffle service for application " +
+          "updateLocalDirsTwiceWithTwoAttempts", re.getMessage());
+      throw re;
+    }
+  }
+
+  @Test(expected = NullPointerException.class)
+  public void testPushBlockStreamCallBackWhileNewAttemptRegistered()
+    throws IOException, InterruptedException {
+    Semaphore closed = new Semaphore(0);
+    pushResolver = new RemoteBlockPushResolver(conf) {
+      @Override
+      void closeAndDeletePartitionFilesIfNeeded(
+        AppShuffleInfo appShuffleInfo,
+        boolean cleanupLocalDirs) {
+        super.closeAndDeletePartitionFilesIfNeeded(appShuffleInfo, cleanupLocalDirs);
+        closed.release();
+      }
+    };
+    String testApp = "testPushBlockStreamCallBackWhileNewAttemptRegisters";
+    Path[] attempt1LocalDirs = createLocalDirs(1);
+    registerExecutor(testApp,
+      prepareLocalDirs(attempt1LocalDirs, MERGE_DIRECTORY + "_" + ATTEMPT_ID_1),
+      MERGE_DIRECTORY_META_1);
+    ByteBuffer[] blocks = new ByteBuffer[]{
+      ByteBuffer.wrap(new byte[4]),
+      ByteBuffer.wrap(new byte[5]),
+      ByteBuffer.wrap(new byte[6]),
+      ByteBuffer.wrap(new byte[7])
+    };
+    StreamCallbackWithID stream1 = pushResolver.receiveBlockDataAsStream(
+      new PushBlockStream(testApp, 1, 0, 0, 0, 0));
+    // The onData callback should be called 4 times here before the onComplete callback. But a
+    // register executor message arrives in shuffle service after the 2nd onData callback. The 3rd
+    // onData callback should all throw NullPointerException as their channels are set to null.
+    stream1.onData(stream1.getID(), blocks[0]);
+    stream1.onData(stream1.getID(), blocks[1]);
+    Path[] attempt2LocalDirs = createLocalDirs(2);
+    registerExecutor(testApp,
+      prepareLocalDirs(attempt2LocalDirs, MERGE_DIRECTORY + "_" + ATTEMPT_ID_2),
+      MERGE_DIRECTORY_META_2);
+    closed.acquire();
+    // Should throw NullPointerException here.
```

Comment:
nit: can we throw a more meaningful exception instead of NPE?

Reply:
This is a good suggestion. But right now in PushBlockStreamCallBack, the close will set dataChannel, indexFile and metaFile to Null, and these three variables are used directly in lots of places. If we need to wrap up this NPE, I think we can do something like this:
```
    @VisibleForTesting
    MergeShuffleFile getIndexFile() {
      if (indexFile == null) {
        throw new RuntimeException(
          String.format("The merged shuffle partition info for appId %s shuffleId %s "
          + "reduceId %s has been cleaned up", appId, shuffleId, reduceId));
      }
      return indexFile;
    }

    @VisibleForTesting
    MergeShuffleFile getMetaFile() {
      if (metaFile == null) {
        throw new RuntimeException(
          String.format("The merged shuffle partition info for appId %s shuffleId %s "
            + "reduceId %s has been cleaned up", appId, shuffleId, reduceId));
      }
      return metaFile;
    }

    @VisibleForTesting
    FileChannel getDataChannel() {
      if (dataChannel == null) {
        throw new RuntimeException(
          String.format("The merged shuffle partition info for appId %s shuffleId %s "
            + "reduceId %s has been cleaned up", appId, shuffleId, reduceId));
      }
      return dataChannel;
    }
```
And update the direct usage of these three variables to call each get method and so it will evaluate the Null values.

But even with this check in get method, it can still be set to Null after the get method returned, then NPE will still be thrown in this case. 

Any other suggestions that we can wrap the NPE for these three variables?

Reply:
How about we just simply catch the NPE and throw the RuntimeException instead?

Reply:
BTW, I'm also thinking that this's not a big issue since the error is replied to previous application attempts and users may care more about the lastest attempt. We can leave it for now if it's not easy to do and improve it later.

Reply:
The NPE is coming from precondition validation failing - not actual NPE due to dereferencing a null, right ?

Reply:
No, it's from the reference:
https://github.com/apache/spark/blob/a97885bb2c81563a18c99df233fb9e99ff368c9c/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java#L487


(`dataChannel` is null in this case)

Reply:
Ah !
Can pull that as a local variable in that method, check nonNull and then use it within the method ?

Reply:
I added a catch clause in the onData and onComplete where they possibly will get NPE when calling the method `writeBuf` and `partitionInfo.updateChunkInfo `and throw RuntimeException, this should fix it.

Reply:
This change was reverted.